cmd/utils/nodecmd/snapshot.go

```diff
@@ -0,0 +1,330 @@
+	if root == (common.Hash{}) {
+		// Corrupt or empty database, init from scratch
+		return errors.New("empty root")
+	}
+
+	logger.Info("Trace Start", "BlockNum", blockNumber)
+
+	sdb, err := state.New(root, state.NewDatabase(dbm), nil)
+	if err != nil {
+		return fmt.Errorf("Failed to open newDB trie : %v", err)
+	}
+	tridb := sdb.Database().TrieDB()
+
+	// Get root-node childrens to create goroutine by number of childrens
+	children, err := tridb.NodeChildren(root)
+	if err != nil {
+		return fmt.Errorf("Fail get childrens of root : %v", err)
+	}
+
+	var wait sync.WaitGroup
+	wait.Add(len(children))
+	// Create goroutine by number of childrens
+	for _, child := range children {
+		go func(child common.Hash) {
+			defer wait.Done()
+			doTraceTrie(sdb.Database(), child)
+		}(child)
+	}
+
+	wait.Wait()
+	logger.Info("Trie Tracer finished")
+	return nil
+}
+
+func doTraceTrie(db state.Database, root common.Hash) (resultErr error) {
+	once.Do(func() {
+		go func() {
+			for {
+				time.Sleep(time.Second * 5)
+				logger.Info("Trie Tracer", "AccNode", midAccountCnt, "AccLeaf", leafAccountCnt, "StrgNode", midStorageCnt, "StrgLeaf", leafStorageCnt, "Unknown", unknownCnt, "CodeAcc", codeCnt)
+			}
+		}()
+	})
+
+	logger.Info("Trie Tracer Start", "Hash Root", root)
+	// Create and iterate a state trie rooted in a sub-node
+	oldState, err := state.New(root, db, nil)
+	if err != nil {
+		logger.Error("can not open trie DB", err.Error())
+		panic(err)
+	}
+
+	oldIt := state.NewNodeIterator(oldState)
+
+	for oldIt.Next() {
+		mutex.Lock()
+		switch oldIt.Type {
+		case "state":
+			midAccountCnt++
+		case "storage":
+			midStorageCnt++
+		case "code":
+			codeCnt++
+		case "state_leaf":
+			leafAccountCnt++
+		case "storage_leaf":
+			leafStorageCnt++
+		default:
+			unknownCnt++
+		}
+		mutex.Unlock()
+	}
+	if oldIt.Error != nil {
+		logger.Error("Error Finished", "Root Hash", root, "Message", oldIt.Error)
+	}
+	logger.Info("Trie Tracer Finished", "Root Hash", root, "AccNode", midAccountCnt, "AccLeaf", leafAccountCnt, "StrgNode", midStorageCnt, "StrgLeaf", leafStorageCnt, "Unknown", unknownCnt, "CodeAcc", codeCnt)
+	return nil
+}
+
+func iterateTrie(ctx *cli.Context) error {
+	stack := MakeFullNode(ctx)
+	dbm := stack.OpenDatabase(getConfig(ctx))
+	sdb, err := state.New(common.Hash{}, state.NewDatabase(dbm), nil)
+	if err != nil {
+		return fmt.Errorf("Failed to open newDB trie : %v", err)
+	}
+
+	logger.Info("TrieDB Iterator Start", "node count : all node count, nil node count : key or value is nil node count")
+	cnt, nilCnt := uint64(0), uint64(0)
+	go func() {
+		for {
+			time.Sleep(time.Second * 5)
+			logger.Info("TrieDB Iterator", "node count", cnt, "nil node count", nilCnt)
+		}
+	}()
+
+	it := sdb.Database().TrieDB().DiskDB().GetStateTrieDB().NewIterator(nil, nil)
+	for it.Next() {
+		cnt++
+		if it.Key() == nil || it.Value() == nil || bytes.Equal(it.Key(), []byte("")) || bytes.Equal(it.Value(), []byte("")) {
+			nilCnt++
+		}
+	}
+	it.Release()
```

Comment:
In other code references, Iterator uses defer Release(). How about moving the 'defer it.Release()' statement to appear right after the 'it' variable is assigned.